falcon/util/structures.py

```diff
@@ -107,6 +107,126 @@ def __repr__(self):
         return '%s(%r)' % (self.__class__.__name__, dict(self.items()))
 
 
+class Context(object):
+    """
+    Convenience class to hold contextual information in its attributes.
+
+    This class is used as the default :class:`~.Request` and :class:`~Response`
+    context type (see
+    :attr:`Request.context_type <falcon.Request.context_type>` and
+    :attr:`Response.context_type <falcon.Response.context_type>`,
+    respectively).
+
+    In Falcon versions prior to 2.0, the default context type was ``dict``. To
+    ease the migration to attribute-based context object approach, this class
+    also implements mapping interface, that is, object attributes are linked to
+    dictionary items, and vice versa, for instance:
+
+    >>> context = falcon.Context()
+    >>> context.cache_strategy = 'lru'
+    >>> context.get('cache_strategy')
+    'lru'
+    >>> 'cache_strategy' in context
+    True
+
+    Note:
+        Python 2 specific ``dict`` methods are exposed regardless of the
+        Python language version, however, as they are delegated to the
+        underlying ``__dict__``, a similar error would be raised as if
+        attempting to use these methods for a usual Python 3 dict.
+    """
+
+    def __contains__(self, key):
+        return self.__dict__.__contains__(key)
+
+    def __delitem__(self, key):
+        self.__dict__.__delitem__(key)
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self.__dict__.__eq__(other.__dict__)
+        return self.__dict__.__eq__(other)
+
+    def __getitem__(self, key):
+        return self.__dict__.__getitem__(key)
+
+    def __hash__(self):
+        return hash(self.__dict__)
+
+    def __iter__(self):
+        return self.__dict__.__iter__()
+
+    def __len__(self):
+        return self.__dict__.__len__()
+
+    def __ne__(self, other):
+        if isinstance(other, type(self)):
+            return self.__dict__.__ne__(other.__dict__)
+        return self.__dict__.__ne__(other)
+
+    def __repr__(self):
+        return '{}({})'.format(type(self).__name__, self.__dict__.__repr__())
+
+    def __setitem__(self, key, value):
```

Comment:
Nitpick: Group related methods in the implementation (e.g., colocate `__getitem__`,  `__setitem__`,  `__delitem__`).